lnst/Tests/PktGen.py

```diff
@@ -0,0 +1,180 @@
+        match = re.search(
+            r"(\d+).+ usec,\s(\d+)\s\(.+,(\d+)frags\).+?.+errors:\s?(\d+)",
+            results,
+            re.DOTALL,
+        )
+        if not match:
+            raise TestModuleError("Could not parse pktgen device result")
+
+        keys = ["duration", "packets", "frags", "errors"]
+
+        try:
+            results = dict(  # converts iterator to actual dict
+                zip(  # zips values with `keys`
+                    keys, map(int, match.groups())  # map converts str numbers to int
+                )
+            )
+        except ValueError:  # just to be sure all the nums can be converted to int
+            raise TestModuleError("Could not convert numbers in pktgen results")
+
+        results["duration"] /= 1_000_000  # from usec to sec
+        return results
+
+
+class PktGen(BaseTestModule):
+    """
+    In the scope of this module, the physical interface is refered as `interface`.
+    Pktgen device (interface@anything) is refered as device.
+
+    Inspired by https://github.com/torvalds/linux/blob/master/samples/pktgen/pktgen_sample03_burst_single_flow.sh
+    """
+
+    cpus = ListParam(type=IntParam())  # each CPU is 1 generator
+
+    pgctrl = StrParam(mandatory=False, default="/proc/net/pktgen/pgctrl")
+
+    src_if = DeviceParam()
+    dst_if = DeviceParam()
+
+    count = IntParam(mandatory=False, default=0)  # 0 = no upper limit
+    size = IntParam(mandatory=False, default=60)  # 4 bytes are added for CRC by NIC
+    frags = IntParam(mandatory=False, default=1)
+    burst = IntParam(mandatory=False, default=8)
+
+    duration = IntParam(default=60)
+
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+
+        self._devices = []
+        self._ip6 = True
+        if isinstance(self.params.dst_if.ips[0], Ip4Address):
+            self._ip6 = False
+
+        self._res_data = {}
+        self._output_parser = None
+
+    def run(self):
+        if not kmod_in_use("pktgen"):
+            raise TestModuleError("pktgen module is not loaded")
+
+        logging.debug("Configuring PktGen generator")
+        self._pg_ctl("reset")
+        self._configure_generator()
+
+        output_parser = PktGenResultsParser(self._devices, self._res_data)
+
+        logging.debug("Starting generator")
+        pktgen = Popen("echo 'start' > /proc/net/pktgen/pgctrl", shell=True)
+        # ^^ echoing start to controller is blocking => needs to be separated
+
+        time.sleep(self.params.duration)
+
+        pktgen.kill()  # stops pktgen
+
+        output_parser.parse_dev_outputs()
+
+        self._deconfigure_generator()
+        return True
+
+    def _configure_generator(self):
+        logging.debug("Configuring generator")
+        src = f"src{6 if self._ip6 else ''}"
+        dest = f"dst{6 if self._ip6 else ''}"
+
+        src_ip = self.params.src_if.ips[0]  # xdp uses just 1 IP
+        dst_ip = self.params.dst_if.ips[0]
```

Comment:
i'm not sure this `[0]` selection will work with our enrt recipes well. in all of them we always configure both the ipv4 as well as an ipv6 address. Their order MAY (though it shouldn't) switch non deterministically... and... both will always be present, so maybe we should write this in a way that doesn't assume that the correct ip address is in the correct place all of the time.

Reply:
not relevant any more